strawberryfields/tdm/tdmprogram.py

```diff
@@ -605,39 +641,52 @@ def unroll(self, shots):
         # q[sm[2]] as concurrent modes of spatial mode C
         # q[sm[3]] as concurrent modes of spatial mode D.
 
+        # save previous mode index of a command to be able to check when modes
+        # are looped back to the start (not allowed when space-unrolling)
+        last_idx = dict()
+
         for cmd in self.rolled_circuit:
+            last_idx[cmd] = 0
             if isinstance(cmd.op, ops.Measurement):
                 self.measured_modes.append(cmd.reg[0].ind)
+
         for i in range(self.timebins):
             for cmd in self.rolled_circuit:
-                self.apply_op(cmd, q, i)
-
-            if self.shift == "default":
+                modes = get_modes(cmd, q)
+                if not (self._is_space_unrolled and any(m.ind < last_idx[cmd] for m in modes)):
```

Comment:
Would consider:
1) breaking up the boolean expression such that `not` acts on it using De Morgan's laws;
2) naming the right expression such that it will be more readable.